pkg/ruler/ruler_sync_queue.go

```diff
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ruler
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	defaultRulerSyncPollFrequency = 10 * time.Second
+)
+
+type rulerSyncQueue struct {
+	services.Service
+
+	queueMx sync.Mutex
+	queue   []string
+
+	pollChan      chan []string
+	pollFrequency time.Duration
+
+	stopped chan struct{}
+}
+
+func newRulerSyncQueue(pollFrequency time.Duration) *rulerSyncQueue {
+	q := &rulerSyncQueue{
+		pollChan:      make(chan []string),
+		pollFrequency: pollFrequency,
+		stopped:       make(chan struct{}, 1),
+	}
+
+	q.Service = services.NewIdleService(q.starting, q.stopping)
+	return q
+}
+
+func (q *rulerSyncQueue) starting(_ context.Context) error {
+	go func() {
+		for {
+			q.queueMx.Lock()
+			userIDs := q.queue
+			q.queue = nil
+			q.queueMx.Unlock()
+
+			if len(userIDs) > 0 {
+				select {
+				case q.pollChan <- userIDs:
+				case <-q.stopped:
+					// We're done.
+					return
+				}
+			}
+
+			// Wait.
+			select {
+			case <-time.After(q.pollFrequency):
+			case <-q.stopped:
+				// We're done.
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (q *rulerSyncQueue) stopping(_ error) error {
+	close(q.stopped)
+	return nil
+}
+
+// enqueue adds to the queue the request to sync rules for the input userID.
+func (q *rulerSyncQueue) enqueue(userID string) {
+	q.queueMx.Lock()
+	defer q.queueMx.Unlock()
+
+	if !slices.Contains(q.queue, userID) {
```

Comment:
If queue is full of tenants already, this can get slow. Shall we use map to reduce that risk?

Reply:
Why not, done: https://github.com/grafana/mimir/pull/4975/commits/c7078b153cf360b08aae7b71f22bacdf3c04d4ed